### PR TITLE
Restructure modules

### LINKED ISSUE #64: Restructure `prif` modules
FILL IN LATER

### PR DESCRIPTION
**Pull request by ktras**
Closes #64.

### CODE REVIEW

File: src/caffeine/collective_helpers_m.f90
Comment (bonachea) [inline src/caffeine/collective_helpers_m.f90:1]: IMO this module is nasty. It's not guaranteed by the PRIF spec, and therefore should not be provided/exported by Caffeine to the client / test code as it currently is.

IIUC it's also unnecessary in test code, which appears could trivially be rewritten to stop using it, allowing this to remain a private submodule. Longer term the `prif_co_reduce` implementation could/should be reworked to stop using this ugly hack entirely (which I believe is also a prerequisite to supporting non-blocking `co_reduce`).


Please enter an issue to clean this up, and we can discuss details in our next meeting.

**Review by bonachea — CHANGES_REQUESTED** [review]
Overall this looks like the wonderful and long-overdue cleanup we discussed.

Added a few questions and minor requests.

File: src/prif.f90 [hunk 1]
```diff
@@ -1,25 +1,655 @@
...
+  public :: prif_form_team, prif_change_team, prif_end_team, prif_team_type, prif_get_team, prif_team_number
+  public :: prif_sync_all, prif_sync_images, prif_sync_team, prif_sync_memory
+  public :: prif_lock, prif_unlock
+  public :: prif_critical, prif_end_critical
+  public :: prif_event_post, prif_event_wait, prif_event_query
+  public :: prif_notify_wait
+  public :: prif_atomic_add, prif_atomic_and, prif_atomic_or, prif_atomic_xor, prif_atomic_cas, prif_atomic_fetch_add
+  public :: prif_atomic_fetch_and, prif_atomic_fetch_or, prif_atomic_fetch_xor, prif_atomic_define, prif_atomic_ref
+
+  type, bind(C) :: handle_data
```
[the remark this answers is not in the document]
Comment (ktras) [reply src/prif.f90]: Thanks for the feedback. The change has been made.

File: src/prif.f90 [hunk 2]
```diff
@@ -1,25 +1,655 @@
...
+    type(c_ptr) :: coarray_data
+    integer(c_int) :: corank
+    integer(c_size_t) :: coarray_size
+    integer(c_size_t) :: element_length
+    type(c_funptr) :: final_func
+    type(c_ptr) :: previous_handle, next_handle
+    integer(c_intmax_t) :: lcobounds(15), ucobounds(15)
+  end type
+
+  type :: prif_coarray_handle
```
[the remark this answers is not in the document]
Comment (ktras) [reply src/prif.f90]: Change made.

File: src/prif.f90 [hunk 3]
```diff
@@ -1,25 +1,655 @@
...
+    private
+    type(handle_data), pointer :: info
+  end type
+
+  type :: prif_team_type
```
[the remark this answers is not in the document]
Comment (ktras) [reply src/prif.f90]: Change made.

File: src/caffeine/prif_private_s.f90 [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+! Terms of use are as specified in LICENSE.txt
+submodule(prif) prif_private_s
+  implicit none
+
+  type(prif_team_type), target :: default_team
+  type(prif_team_type), pointer :: current_team => null()
+  type(c_ptr) :: non_symmetric_heap_mspace
+
+  ! TODO: Do these types need to be public? They are not used as a type of any of the args
+  ! for the public procedures
```
[the remark this answers is not in the document]
Comment (ktras) [reply src/caffeine/prif_private_s.f90]: Of course, thanks for the reminder as to why they must be public. Change made.

File: src/prif.f90
[the remark this answers is not in the document]
Comment (ktras) [reply src/prif.f90:1]: I made no changes to the contents when moving them. Just cut and paste from one module to the other.

**Comment by ktras** [comment]
I just moved the definition of `handle_data` to the bottom of the `prif` module. `prif_coarray_handle` and `prif_team_type` type definitions cannot be moved to the same location because there are arguments of those types above.

**Comment by bonachea** [comment]
> I just moved the definition of `handle_data` to the bottom of the `prif` module. `prif_coarray_handle` and `prif_team_type` type definitions cannot be moved to the same location because there are arguments of those types above.

I thought we tried this in our prototyping meeting and decided that at least one compiler rejected the declaration of `prif_coarray_handle` when the `handle_data` pointer field was an incomplete type? What compilers have you successfully tested against?

Of course what ultimately matters most is what flang-new will accept, which I'm guessing is currently "none of this". On a related note, I think this PR needs to update the table in the top-level README where we track major modern Fortran features the Caffeine implementation requires. We at least need to add "Submodule support [1]" (which I believe was added in F08?), and probably remove the [1] from `contiguous` attribute, which is now "baked into" the PRIF API. There may be other updates needed.

**Review by bonachea — CHANGES_REQUESTED** [review]
Thanks for the recent changes, which look good.

Some unresolved tasks remain.

**Comment by ktras** [comment]
> I thought we tried this in our prototyping meeting and decided that at least one compiler rejected the declaration of `prif_coarray_handle` when the `handle_data` pointer field was an incomplete type? What compilers have you successfully tested against?

When compiling the whole software stack, the only compiler we are currently developing with is gfortran. 

Based on this comment, I tried compiling just the `prif` module (with the `handle_data` type definition at the bottom of the module) with the following compilers and with the following results:

|   Compiler  |  Result  | 
|------------|---------|
| gfortran      |     compiles        | 
| ifx                |     compiles        | 
| cray ftn       |     compiles        | 
| nvfortran    |     compiles        | 
| nagfor        |  doesn't compile for different reason (doesn't have `c_ptrdiff_t` support). |

**Comment by ktras** [comment]
@bonachea Thank you for the further feedback. I have answered your compiler question above and then updated the PR in reference to your requested changes around `collective_helpers_m` and replied to that comment above. I believe I have addressed all of your review comments to this point. Please let me know about any further feedback, thanks!

**Review by bonachea — APPROVED** [review]
LGTM! Thanks for the overhaul and recent fixes.

**Comment by bonachea** [comment]
Wooops, I just noticed that doc-generator.md also needs updates as a result of this reorg, assuming we are still maintaining the FORD docs.

File: src/caffeine/prif_private_s.f90 [hunk 5]
```diff
@@ -0,0 +1,10 @@
+! Copyright (c), The Regents of the University of California
+! Terms of use are as specified in LICENSE.txt
+submodule(prif) prif_private_s
+  implicit none
```
Comment (bonachea) [inline src/caffeine/prif_private_s.f90:4]: The Fortran spec rules seem potentially ambiguous regarding the default accessibility behavior of a submodule.

Let's add an explicit private here, both to ensure all compilers treat this submodule's data as private to the `prif` implementation and for documentation purposes.

```suggestion
  implicit none
  private
```

**Review by bonachea — CHANGES_REQUESTED** [review]
After additional thought, I've added two further minor requests.

File: src/caffeine/prif_private_s.f90 (hunk 5 above)
Comment (ktras) [reply src/caffeine/prif_private_s.f90:4]: @bonachea Section `8.5.2 Accessibility attribute` of the Fortran 2023 Standard has a constraint that disallows this.
`C817 An access-spec shall appear only in the specification-part of a module.`

Comment (everythingfunctional) [reply src/caffeine/prif_private_s.f90:4]: > The Fortran spec rules seem potentially ambiguous regarding the default accessibility behavior of a submodule.

Only entities declared in a module are accessible via a `use` of that module. All submodule contents are private implementation details.

**Review by everythingfunctional — APPROVED** [review]
LGTM, especially since the changes to the tests were so minimal.

File: src/caffeine/prif_private_s.f90 (hunk 5 above)
Comment (bonachea) [reply src/caffeine/prif_private_s.f90:4]: > Only entities declared in a module are accessible via a `use` of that module. All submodule contents are private implementation details.

Thanks for the confirmation Brad. I suspected this was the case, but could not find confirmation on a brief skim of the relevant spec sections.

**Review by bonachea — APPROVED** [review]
LGTM! Thanks for this valuable cleanup

**Pull request merged**